feature: DeepmaintRooms

scrdest opened the pull request:
Adds DeepmaintRooms, an alternate take on Deepmaint using WaveFunction Collapse implemented in a Rust DLL for mapgen.

![deepmaint-showoff](https://github.com/UristMcStation/UristMcStation/assets/4363639/e98159c1-80c0-4d03-97ac-de0a5e789ac9)

[DeepmaintDemo1.webm](https://github.com/UristMcStation/UristMcStation/assets/4363639/f414acf6-990d-4c43-8829-c4e360eb4d00)


Currently only spawnable/accessible through adminbus measures.

### Features:

- A liminal horror take on Deepmaint - very rare, if any, face-eating monsters... but plenty of non-Euclidean geometries
- Fast mapgen; the Rust side completes in _miliseconds_ for a 90-tile map, the slight stutter occurs due to BYOND-side integrations
- The generated map is duplicated into four _nearly but not exactly_ identical variants. Variants are connected by invisible, silent, randomly-triggered teleporters, creating the illusion of subtle reality shifts.
- Reality shifts are the main threat in DeepmaintRooms - if you're unlucky, a shift may wall you in.
- Death in Deepmaint returns your dead body back to the station; in case of gibbing, spawns a gibspawner with your DNA instead.
- My best effort at neat modularisation of the added code; all integration can be turned off by unticking one file.

### Mapgen

The map is generated by a Rust DLL written by yours truly and implementing a WaveFunction Collapse algorithm. Hadn't pushed the full source code for the version used here, but it's a slight modification of this branch: https://github.com/scrdest/MorkovChain-maptiler/tree/feature/singlethreaded plus the `byond_fn` crate for integration into, well, BYOND.

The generator is codebase-agnostic and configurable; you simply point it at a JSON ruleset that defines:
- **arbitrary numerical keys** 
- (un-normalized, optionally-direction-specific) Markovian **transition weights** between them
-  **map size** (but see below) 
- optionally, an **adjacency class** (`octile` - all adjacents or `cardinal` - no diagonals, only straight-line neighbors; the former requires either specifying more rules or using undirected weights). 

Missing transition weights are interpreted as zeros.

The interpretation of the map into SS13 is done on the BYOND side, with the keys currently interpreted as pointing to a 3x3 'prefab' datum defining a 'patch' of tiles. This means that a size 30 in the mapgen ruleset becomes a 90x90 map in SS13, a size 20 map becomes a 60x60, and so on.

The generator can be reused with different rulesets for different applications; for example, mapping 1 mapgen tile to 1 SS13 turf with a particular undirected ruleset yields nice planets with plausible-looking coastlines.

### TODOs:

- [x] Remove debug code
- [x] Tooling to teleport people into Deepmaint
- [x] Event to trigger Deepmaintey anomalies to pull people wandering maint in
- [ ] Maybe add some extra junk for variety's sake
- [x] Possibly optimize the generation procs on the DM side to not have to do a locate() over and over or rely on hand-entered Z-level but use caching instead
- [ ] Long-term, extend the mapgen library to be able to extend existing maps nicely or freeze portions and regenerate the rest (e.g. for regenerating only the tiles not visible to the player at the moment)

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You'll find a README and example file in .\html\changelogs\ for further instructions.

You can also find a template for adding your changelog directly to the PR description here: https://github.com/Baystation12/Baystation12/wiki/Automatic-changelog-generation
-->

Commit:
feature: Adds DeepmaintRooms, an alternate take on Deepmaint using WaveFunction Collapse implemented in a Rust DLL for mapgen

Currently only spawnable/accessible through adminbus measures

Commit:
chore: Nuked all the "<<" I could find. Bumped expected spawn()s, because they are essential to the proper functioning here.

Commit:
fix: Removes the fallback to civvie areas on deepmaint death hook

This area is only defined for Nerva so it broke on non-Nerva tests and the usage was only a backup solution anyway

Commit:
fix: Unit-tests shouldn't fail the hook

Commit:
fix: Fixes the right thing lol

Commit:
feature: Removed debug stuff from compilation; Admin tools now provided to generate map and throw people in. Better mapgen, less dead ends. Better, spoopier lights. Replaced most object teleporters with turf proc magic for performance.

Commit:
fix: CI stuff

scrdest commented on code/modules/urist/admin/verbs.dm:
Forgot to mention that in the commit message; these changes are meant to optimize those operations by turning them asynchronous. We never cared about the return value here, so should behave exactly the same.

CPU-Blanc commented on code/modules/urist/modules/wfc/core/calllib.dm:
```diff
@@ -0,0 +1,99 @@
+/proc/generate_wfc_map_full(var/rules_json = "deepmaint.json", var/overwrite_all = FALSE, var/zlevel = 1)
+	var/mapname = "genmap.json"
+	generate_wfc_map(rules_json)
+	generate_from_wfc_file(mapname, overwrite_all = FALSE, zlevel = zlevel)
```
Is this necessary? The argument name matches the parameter name and we're not skipping over any.. Seems a little odd to me
```suggestion
	generate_from_wfc_file(mapname, overwrite_all = FALSE, zlevel)
```

scrdest replied:
I generally prefer named arguments, especially for optional params - not only is it clearer, it's also robust to refactoring changing the order of optional args.

Also, I'm used to Python screaming at you for putting positionals after keyword args.

CPU-Blanc commented on code/modules/urist/modules/wfc/_module.dm:
```diff
@@ -0,0 +1,32 @@
+#endif
+
+#include "core\calllib.dm"
+//#include "core\debug.dm"
```
Could be an idea to put this behind some kind of debug `#ifdef`? Along with more verbose logging, like we did in the GOAI project?

scrdest replied:
I think that's redundant in this case, if you want these you can just uncomment the line or manually include debug. It's mostly debug _verbs_ anyway.

CPU-Blanc commented on code/modules/urist/modules/wfc/core/turf.dm:
```diff
@@ -0,0 +1,40 @@
+
+
+/proc/WfcChangeTurf(var/turf/trg, var/new_type)
+	if(isnull(new_type))
+		return
+
+	if(!ispath(new_type))
+		return
+
+	if(!ispath(new_type, /turf))
+		return
```
Seems like some redundant checks here. `ispath(new_type, /turf)` would return false if it is not a path, or if it were null.
```suggestion
	if(!ispath(new_type, /turf))
		return
```

scrdest replied:
I left those in as placeholders for more specific debug logging ("oh, it failed because it's a null/not a turf/an instance")

CPU-Blanc commented on code/modules/urist/modules/wfc/core/turf.dm:
```diff
@@ -0,0 +1,40 @@
+		for(var/overwritten in trg.associated_wfc_overwritables)
+			qdel_from_weakref(overwritten)
+
+	trg.icon_state = initial(trg.icon_state)
```
Seems odd we're accounting for a null `trg` above, but not here?

scrdest replied:
Good spot, oversight.

CPU-Blanc commented on code/modules/urist/modules/wfc/core/turf.dm:
```diff
@@ -0,0 +1,40 @@
+	#ifdef STUB_OUT_SS13
+	var/turf/new_turf = new new_type(trg)
+	#else
+	trg.ChangeTurf(new_type)
```
Same as above. No null check but one prior. Is this intentional?

scrdest replied:
Nope :^)

scrdest replied:
Done; we will have checked it's not null, me using the null-conditional was reflexive - but if we do use it needlessly, that's probably some overhead.

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,36 @@
+	set name = "Generate DeepmaintRooms"
+
+	if(!GLOB.wfc_deepmaint_zlevels_by_instance)
+		log_and_message_admins("- no Deepmaint Z-levels registered!")
```
This log is lacking some context and might be out of place for other staff. Would a simple `to_chat` with error span class be more suitable?

scrdest replied:
IIRC the method logs what proc was called by whom; it's a weird API

scrdest replied:
Also this follows the convention of other Template-adjacent stuff IIRC

CPU-Blanc replied:
Normal logs do not include calliing procs, liines etc. Only `WARNING`, `crash_with` and `ARGS_DEBUG` insert the calling proc and line automatically.
`log_and_message_admins` just inserts the calling user/mob with the `(?)` href link

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,36 @@
+		log_and_message_admins("- no Deepmaint Z-levels registered!")
+		return
+
+	for(var/area/map_template/deepmaint_wfc/deepmaint_area in GLOB.wfc_deepmaint_zlevels_by_instance)
+		deepmaint_area.generate_deepmaint("deepmaint.json", "genmap.json")
+		log_and_message_admins("has generated a DeepmaintRooms map.")
+		break
```
Seems a little odd to break on the first iteration without any kind of checks. Could you not just access `wfc_deepmaint_zlevels_by_instance` by the first index instead?
```suggestion
	var/area/map_template/deepmaint_wfc/deepmaint_area = GLOB.wfc_deepmaint_zlevels_by_instance[1]
	deepmaint_area.generate_deepmaint("deepmaint.json", "genmap.json")
	log_and_message_admins("has generated a DeepmaintRooms map.")
```

scrdest replied:
No, because _you're not guaranteed to have an index 1_. The current code handles a failure gracefully and doesn't report success on a failure.

CPU-Blanc replied:
I actually thought about this after I turned the PC off. Fair point!

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,36 @@
+		log_and_message_admins("tried to throw into Deepmaint, but no Deepmaint Z-levels are currently registered!")
+		return
+
+	var/atom/movable/target = M
```
Any reason not to typecast this in the proc parameters instead of reassigning like this? I don't _think_ the admin datum type checks when calling the proc, so would a `istype(M)` be required in place of the `isnull` on the next line?

scrdest replied:
IIRC I did this to extend this to all atoms later; it was just annoying to have this verb pop up for all atoms, for time being.

CPU-Blanc reviewed (commented):
I'm sometimes smoothbrain, but I thought a smoothbrain second pair of eyes are better than nothing?

Just some little bits I spotted, not gone through all of it just yet but it's getting late

scrdest commented on code/modules/urist/admin/verbs.dm:
```diff
@@ -11,13 +12,14 @@
 		M.loc = get_turf(usr)
 		message_admins("[key_name(usr)] has warped all players to their location.")
 		log_admin("[key_name(src)] has warped all players to their location.")
```
Oh, I think you're right; probably part of the reason this is a hella heavy proc.

CPU-Blanc replied:
Also this can just be a single call to` log_and_message_admins`, dropping the `key_name` as it's automatically inserted

scrdest commented on code/modules/urist/modules/wfc/_wfc_defines.dm:
```diff
@@ -0,0 +1,43 @@
+#define DEEPMAINT_FLOOR_ICON 'icons/turf/flooring/plating.dmi'
+#define DEEPMAINT_FLOOR_ICONSTATE "plating"
+
+#define WFC_CHANGE_TURF(From, To) WfcChangeTurf(From, To)
```
It was originally different between the two projects and in the two ifdef branches below. I kept it here as a way of having a zero-cost abstraction over the implementation if things change _again_.

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,143 @@
+		// This would be silly
+		return
+
+	if(istype(trg, /mob) && !istype(trg, DEEPMAINT_TELEPORTABLE_MOBTYPE))
```
Not sure what DM does under the hood, but I'd hope using a built-in proc would be better for this?
```suggestion
	if(ismob(trg) && !istype(trg, DEEPMAINT_TELEPORTABLE_MOBTYPE))
```

scrdest replied:
I didn't realize this was an actual builtin, since we have a couple of macros that pretend to be builtins.

CPU-Blanc replied:
Yeah. All the _basic_ datums have built-in procs: `ismob()`, `isobj()`, `isturf()`, `isarea()` and also `ismovable()`. As you said, there _are_ a load that are just `istype` checks wearing coats, like `isAI()`

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,143 @@
+
+
+/proc/deepmaint_send_trg(var/atom/movable/trg)
+	if(!istype(trg, /atom/movable))
+		return FALSE
+
+	if(isnull(trg.loc))
+		// No un-qdel-ing by accident!
+		return
+
+	if(istype(trg, /turf) || istype(trg, /area))
+		// This would be silly
+		return
+
+	if(istype(trg, /mob) && !istype(trg, DEEPMAINT_TELEPORTABLE_MOBTYPE))
+		// Don't teleport AI eyes, ghosts, irate admins...
+		return
```
Same as above comments

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,143 @@
+
+
+/proc/deepmaint_return_trg(var/atom/movable/trg)
+	if(!istype(trg, /atom/movable))
+		return FALSE
+
+	if(isnull(trg.loc))
+		// No un-qdel-ing by accident!
+		return
+
+	if(istype(trg, /turf) || istype(trg, /area))
+		// This would be silly
+		return
+
+	if(istype(trg, /mob) && !istype(trg, DEEPMAINT_TELEPORTABLE_MOBTYPE))
+		// Don't teleport AI eyes, ghosts, irate admins...
+		return
```
As above

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/turfs.dm:
```diff
@@ -0,0 +1,146 @@
+			teleported = deepmaint_conditional_stationyeet(Obj, tele_proba = src.wfc_teleproba_station)
+
+		if(!teleported && src.wfc_has_variantlink)
+			teleported = deepmaint_conditional_yeet(Obj, tele_proba = src.wfc_teleproba_deepmaint)
```
Updating `teleported` here doesn't seem to do anything as it's immediately discarded. Is this intentional?

scrdest replied:
Yeah; maintainability. This way you can swap the priority of the two checks, add prob() and whatever, and it will still work. 

The intended behavior is that the first successful teleport prevents any further teleports without messing with early returns which can make cleanup logic hard.

CPU-Blanc replied:
Makes sense :+1:

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/turfs.dm:
```diff
@@ -0,0 +1,146 @@
+/turf/deepmaint/wall/probably/New(var/atom/loc, var/improbability_override = null)
+	var/wall_improba = (isnull(improbability_override) ? src.improbability : improbability_override)
+
+	if(prob(wall_improba))
+		spawn(0)
+			WFC_CHANGE_TURF(src, /turf/deepmaint/plating)
```
Will this not cause some init issues? iirc it hates it when stuff are qdel'd before being initialised as a call to parent is never made here. Might be wrong, but I remember having similar issues with landmarks

scrdest replied:
I've ran this, looked fine. 

The trick is, I don't think this _actually_ isn't initialized by the time of qdel(). The reason being that `spawn(0)` - anything behind it gets put at the _end_ of the queue of pending futures to schedule, while negative values prepend it at the _front_ of the queue.

So, the scheduler ordering goes `New() -> ...any spawn(-1)'d code... -> ...any prior spawn(0)'d code... ->WFC_CHANGE_TURF() -> ...any spawn(0)'d code after this...`

CPU-Blanc commented on ss13_wfc.dll:
Any thoughts of open-sourcing this dll? Is there a way we could open a new repo under Urist for this/fork it from your repo?

scrdest replied:
Yeah, I'm planning to; the trick is that this a wrapper Rust library I wrote on top of my _second_ Rust library... and then to top it all off, the SS13 bits are on an _un-pushed branch_ of that second library so the first one is pointed at my _local filesystem_ ATM 🤡 

The base library ('second' above) is this: https://crates.io/crates/morkovmap_rc

CPU-Blanc reviewed (commented):
The next batch of smoothbrain comments. Ignore what are irrelevant, but just some general comments I had

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/generation.dm:
```diff
@@ -0,0 +1,47 @@
+	var/mapname = "genmap.json"
+	var/_sleeptime = max(1, sleeptime)
+
+	spawn(0)
+		while(TRUE)
+			deepmaint_rebuild_mapfile(ruleset)
+			deepmaint_from_mapfile(mapname, overwrite_all = FALSE, zlevel = zlevel)
+			to_chat(world, "Regenerating in 20s")
+			sleep(_sleeptime)
+			to_chat(world, "Regenerating!")
```
Also left in the output to `world` :P 
Might want to make this a debug log instead
```suggestion
	spawn(0)
		while(TRUE)
			deepmaint_rebuild_mapfile(ruleset)
			deepmaint_from_mapfile(mapname, overwrite_all = FALSE, zlevel = zlevel)
			log_debug("Deepmaint - Regenerating in 20s")
			sleep(_sleeptime)
			log_debug("Deepmaint - Regenerating!")
```

scrdest commented on code/modules/urist/modules/wfc/modules/deepmaint/hooks.dm:
```diff
@@ -0,0 +1,31 @@
+/hook/death/proc/deepmaint_death_handler(var/mob/living/carbon/human/H, var/gibbed = FALSE)
+	var/mob/living/carbon/human/deadman = H
```
IIRC, DM doesn't _actually_ validate input params, IIRC, so you can wind up runtiming by calling procs on a type that does not have them. 

The cast in L2 here is doing the heavy lifting of nulling the input out if it's not the expected type so that we can do null checks on it to catch both mis-typed and explicitly null inputs.

CPU-Blanc replied:
Yep, DM doesn't validate types passed to parameters. But iirc, assigning like this _doesn't_ null on incorrect types (might require validation, but I'm pretty sure that's how DM works). Pretty sure DM is kinda dumb like that. You'd have to do a `istype(H)` to catch both nulls and incorrect types

scrdest commented on code/modules/urist/modules/wfc/modules/deepmaint/hooks.dm:
```diff
@@ -0,0 +1,31 @@
+
+	else
+		H.forceMove(T)
+		H.visible_message("[H] materializes on [T], seemingly out of thin air.")
```
Yeah, good shout. I don't trust text macros much so I forget about them 😅

scrdest commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,143 @@
+
+/proc/deepmaint_yeet_trg(var/atom/movable/trg, var/xoffset = 1, var/yoffset = 1)
+	if(!istype(trg, /atom/movable))
```
Maybe, but I find it much less readable. You need to mentally disambiguate between the is-implicit-type case and the is-SOME-type case. Because we cannot have sane APIs in BYOND.

CPU-Blanc replied:
Fair. I think it's personal preference. For _me_ `istype(trg)` tells me I'm checking if what I have is what I expected, whereas `istype(trg, /atom/movable)` is checking if it's a specific child type of what we currently have.

CPU-Blanc replied:
_This could also be just replaced with the built-in `ismovable(trg)`_

scrdest commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,143 @@
+	if(!istype(trg, /atom/movable))
+		return FALSE
+
+	if(isnull(trg.loc))
+		// No un-qdel-ing by accident!
+		return
```
Eh, maybe, but this would be more painful to port out of SS13 for standalone testing and probably catch less cases.

CPU-Blanc replied:
Are there cases where a datum might be marked for GC but still have a loc, however? That could be missed here if so

scrdest replied:
Shouldn't be a problem in that case; even if we teleport it, it will be moved into nullspace by the GC, just from a different location. 

The primary purpose of this is to avoid accidentally creating new refs by moving things out of nullspace, whether it's there due to GC or some nonsense (e.g. being inside an object which then gets deleted, but the logic doesn't transitively delete the contents fsr).

scrdest commented on code/modules/urist/modules/wfc/modules/landmass/prefabs.dm:
```diff
@@ -0,0 +1,110 @@
+	var/turf_type = null
+
+	switch(state)
+		if(1)
```
This is just me being lazy honestly... plus this is only ever used in this one proc to remap from the JSON output to prefabs and will never be needed elsewhere. 

The overhead of checking what the macro maps to vs what's in the JSON might well be higher than if we keep it plaintexted - it's already obvious what _prefab_ the ID maps to.

scrdest committed:
Apply suggestions from code review

Co-authored-by: CPU_Blanc <54823378+CPU-Blanc@users.noreply.github.com>

Commit:
tweak: More CR fixes

CPU-Blanc commented on code/modules/urist/admin/verbs.dm:
```diff
@@ -1,23 +1,25 @@
+		M.forceMove(get_turf(usr))
+		sleep(-1)
+
+	log_and_message_admins(" has warped all players to their location.")
```
Minor nit-pick, but I'd personally prefer this log to be _before_ the `for` loop. If there's any lag it'll give a better indication as to what has caused it. Also iirc you don't need that leading space within the string, that's already added in the proc

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,36 @@
+	set name = "Generate DeepmaintRooms"
+
+	if(!GLOB.wfc_deepmaint_zlevels_by_instance)
+		log_and_message_admins("- failed to generate Deepmaint - no Deepmaint Z-levels registered!")
```
Super nit-pick. Leading Hyphen might look a little odd and out of place with the admin log

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/generation.dm:
```diff
@@ -0,0 +1,47 @@
+	var/mapname = "genmap.json"
+	var/_sleeptime = max(1, sleeptime)
+
+	spawn(0)
+		while(TRUE)
+			deepmaint_rebuild_mapfile(ruleset)
+			deepmaint_from_mapfile(mapname, overwrite_all = FALSE, zlevel = zlevel)
+			log_debug("Deepmaint - Regenerating in 20s")
+			sleep(_sleeptime)
+			log_debug("Deepmaint - Regenerating!")
```
I see logging got fixed but anything re: the dangling `while(TRUE)` loop concern? Or is not an issue with how DM handle their wonk async?

scrdest replied:
Sorry - I remembered seeing this comment and I've literally double-checked but couldn't find your original comment, so I thought you found it's fine yourself and deleted it.

I've had a hunch that it's fine, and it is - BYOND being a lovely mess that it is, that fact is hidden in the documentation for `del()`:

```Procs that are executing with src equal to that object are silently killed, causing execution to return to the caller. If that is not what you want, you should detach the proc from the source object by setting src to null.```

Since this is indeed attached to an object (the area), it should be fine.

CPU-Blanc reviewed (commented):
Last few nit-picks/questions from me

CPU-Blanc commented:
Just need to add `/area/map_template/deepmaint_wfc` to `maps/away_sites_testing/template_testing_unit_testing.dm` in the `area_usage_test_exempted_root_areas` list to fix that test fail. Not sure why it didn't fail before and only now..

Commit:
fix: CR/CICD fixes

Commit:
fix: Well the first try didn't work! Maptesting take 2

Commit:
fix: Rebase & adding stuff to the right place. Big smort.

scrdest commented:
@CPU-Blanc I've pushed the source for the libcode I've used for the DLLs here: https://github.com/scrdest/MorkovChain-maptiler/tree/feature/ss13

The actual bindings to DM are in a separate crate, but the source is so extremely tiny I can just specify this inline:
```Rust
use morkovmap_rc;
use byond_fn::byond_fn;

#[byond_fn]
fn wfc() {
    morkovmap_rc::generate_from_file(None)
}

#[byond_fn]
fn from_ruleset(ruleset: &str) {
    morkovmap_rc::generate_from_file(Some(ruleset))
}
```
The only deps are this lib and the byond_fn crate's macro for cajoling the calling convention stuff because I CBA to do it manually.

CPU-Blanc commented:
> @CPU-Blanc I've pushed the source for the libcode I've used for the DLLs here: https://github.com/scrdest/MorkovChain-maptiler/tree/feature/ss13
> 
> The actual bindings to DM are in a separate crate, but the source is so extremely tiny I can just specify this inline:
> 
> ```rust
> use morkovmap_rc;
> use byond_fn::byond_fn;
> 
> #[byond_fn]
> fn wfc() {
>     morkovmap_rc::generate_from_file(None)
> }
> 
> #[byond_fn]
> fn from_ruleset(ruleset: &str) {
>     morkovmap_rc::generate_from_file(Some(ruleset))
> }
> ```
> 
> The only deps are this lib and the byond_fn crate's macro for cajoling the calling convention stuff because I CBA to do it manually.

Haven't had time tonight to look at the source much just yet, but couldn't you make it a workspace, include the wrapper crate in it and make it depend on the wfc crate? That way both are together on the same repo, and you could build either one independently (wrapper would obviously build both)?

Ooor probably better, make this a crate feature and gate it behind that? ie
```toml
#Cargo.toml
[dependencies]
byond_fn = { version = "blaah", optional = true } 

[features]
byond = ["dep:byond_fn"]
```
```rust
//lib.rs
#[cfg(feature = "byond")]
#[byond_fn]
fn wfc() {
    morkovmap_rc::generate_from_file(None)
}

#[cfg(feature = "byond")]
#[byond_fn]
fn from_ruleset(ruleset: &str) {
    morkovmap_rc::generate_from_file(Some(ruleset))
}
```

Not done a lot of Rust packaging and mostly dealt with application work, so I might be totally off-base here :P

scrdest commented:
> Haven't had time tonight to look at the source much just yet, but couldn't you make it a workspace, include the wrapper crate in it and make it depend on the wfc crate? That way both are together on the same repo, and you could build either one independently (wrapper would obviously build both)?

Well, I probably _could_, but even then I wouldn't _want_ to. 

Separation of concerns. The WFC crate is meant to be a very generic Rust library for use by... whoever. I don't want to pollute a general-purpose library with an extremely niche use-case.

The integration to SS13 hasn't even been an afterthought, it has been a random lightbulb moment as I was already starting to just port the code to pure DM.

CPU-Blanc commented:
> > Haven't had time tonight to look at the source much just yet, but couldn't you make it a workspace, include the wrapper crate in it and make it depend on the wfc crate? That way both are together on the same repo, and you could build either one independently (wrapper would obviously build both)?
> 
> Well, I probably _could_, but even then I wouldn't _want_ to.
> 
> Separation of concerns. The WFC crate is meant to be a very generic Rust library for use by... whoever. I don't want to pollute a general-purpose library with an extremely niche use-case.
> 
> The integration to SS13 hasn't even been an afterthought, it has been a random lightbulb moment as I was already starting to just port the code to pure DM.

That's why putting behind a feature gate would work perfectly. Those who don't want the niche byond integration won't  get it, but it's ready to go for those that do. The `byond_fn` dependency won't download unless the feature flag is toggled, and the code won't  be compiled and added to the .dll on build. I don't really see a downside other than the 2KB or so it'd take up in the raw source file?
Just seems really silly having such a tiny wrapper be its own crate imo

scrdest commented:
Eh, I'd rather keep it this way and potentially extend it later on with more complicated WFC use-cases (like conditional generation API to freeze some tiles as open to force path generation or w/e) and non-WFC features for SS13. 

There's a _lot_ of cool shit Rust DLLs could do, especially if we had the 515+ native FFI rather than the shitty string FFI. I have my eyes on a couple of scientific crates, e.g. one that simulates chemical reactions stochastically, for instance. And that's without getting into atmos and other such nonsense.

Commit:
feature: Added a proper Deepmaint event. Generates the z-level if needed, no manual work required.
Fixed broken generation, added a bunch of status-tracking retvals.
Disabled the module-based includes for now because DM is a dumb.

CPU-Blanc commented on baystation12.dme:
```diff
@@ -3488,6 +3488,21 @@
 #include "code\modules\urist\modules\weather\weather_objects.dm"
 #include "code\modules\urist\modules\weather\weather_turf.dm"
+#include "code\modules\urist\modules\wfc\_wfc_defines.dm"
+#include "code\modules\urist\modules\wfc\fractal_coords.dm"
+#include "code\modules\urist\modules\wfc\core\calllib.dm"
+#include "code\modules\urist\modules\wfc\core\maptricks.dm"
+#include "code\modules\urist\modules\wfc\core\prefabs.dm"
```
Not sure what's going on here. But I think your includes have got very mixed up?
Many of these are included in `code/modules/urist/modules/wfc/_module.dm` already, however that file isn't ever included anywhere? Some of these includes are behind `ifdef`s too. Are these all supposed to be in the dme, rather than a single module file which then further includes?

Might need a pass?

scrdest replied:
I left a comment in the module file and the commit message 😅 

TL;DR - DM was acting weird with call()s on code included from the submodule, and this had been hanging around for a long time now so I'd rather get it in and fix the includes separately.

The current `_module.dm` is NOT included to avoid double-including and the tick-included featureset corresponds to the correct SS13 set of includes.

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,42 @@
+		return
+
+	for(var/area/map_template/deepmaint_wfc/deepmaint_area in GLOB.wfc_deepmaint_zlevels_by_instance)
+		var/success = deepmaint_area.generate_deepmaint("deepmaint.json", "genmap.json")
+		if(success)
```
`success` var is never read. Could we not save an allocation?
```suggestion
		if(deepmaint_area.generate_deepmaint("deepmaint.json", "genmap.json"))
```

scrdest replied:
It's a stack value, it shouldn't matter - whether you assign it to a var or not, it's _there_ in the memory already (otherwise you couldn't if on it).

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,42 @@
+	for(var/area/map_template/deepmaint_wfc/deepmaint_area in GLOB.wfc_deepmaint_zlevels_by_instance)
+		var/success = deepmaint_area.generate_deepmaint("deepmaint.json", "genmap.json")
+		if(success)
+			log_and_message_admins("has generated a DeepmaintRooms map.")
+			break
+		else
+			log_and_message_admins("has unsuccessfully attempted to generate a DeepmaintRooms map.")
```
Would these not make more sense outside of the loop, to avoid spam in case of concurrent failures? Would also give a reason to use the `success` var defined in the previous comment

scrdest replied:
Fair shout, but as-is it gives you an indication of how badly messed up it is (normally, it should work on first try; if you see multiple failures, something has gone off the rails.

CPU-Blanc replied:
Eh, do we really need to spam all admins with it though? Each call is also an I/O write to file iirc. Could you not make a counter var if you want to see how badly it screwed up, and count the amount of attempts and log that instead?

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,42 @@
+			break
+		else
+			log_and_message_admins("has unsuccessfully attempted to generate a DeepmaintRooms map.")
+
+	return
+
```
Redundant
```suggestion
```

scrdest replied:
Yes, but I hate implicit returns; you have to think more to guess the return value and it makes refactoring uglier if a diff accidentally joins two procs together.

CPU-Blanc replied:
Eh. I personally stick to whatever code style the rest of the code uses, which in this case is implicit returns if you're not returning a var directly. I'm not fussed too much either way, but it's not really done anywhere else so stands out.

I believe this has been addressed in 515 though. iirc you can, much in the style of python, define a return type for a proc.

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,42 @@
+	set name = "Send to DeepmaintRooms"
+
+	if(!GLOB.wfc_deepmaint_zlevels_by_instance)
+		log_and_message_admins("tried to throw [M] into Deepmaint, but no Deepmaint Z-levels are currently registered!")
```
Text macro might make sense here, as it's for general mobs?
```suggestion
		log_and_message_admins("tried to throw \a [M] into Deepmaint, but no Deepmaint Z-levels are currently registered!")
```

scrdest replied:
Sure, why not?

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/admins.dm:
```diff
@@ -0,0 +1,42 @@
+	if(isnull(target))
+		return
+
+	var/success = deepmaint_send_trg(target)
+	if(success)
```
Again, unnecessary allocation
```suggestion
	if(deepmaint_send_trg(target))
```

scrdest replied:
As above, TL;DR suggestion does not save memory and makes refactoring later harder.

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/generation.dm:
```diff
@@ -0,0 +1,51 @@
+/proc/deepmaint_rebuild_mapfile(var/ruleset = "deepmaint.json")
+	var/success = generate_wfc_map(ruleset)
+	return success
+
+/area/map_template/deepmaint_wfc/proc/deepmaint_rebuild_mapfile(var/ruleset = "deepmaint.json")
+	var/success = generate_wfc_map(ruleset)
+	return success
```
Are these procs actually necessary at all? `generate_wfc_map()` already returns a boolean, so these wrapper procs are adding nothing at all other than a name change. Can we not just call `generate_wfc_map(ruleset)` where necessary instead as there's no extra logic here?

scrdest replied:
Not strictly needed, no. They are largely a legacy of a weird bug I've seen with the module-style includes.

For some reason callproc would flat out ignore 'functional' procs from module-included code but would recognize 'method' procs from it... so I bound it to the area.

Partially, I've also kept it in as an OOP-ish API into the underlying functional API, pytorch-style.

CPU-Blanc replied:
iirc DM _sucks_ when it comes to proc calls and their related overhead. If you want to keep them that's fine, but for me it just feels like unnecessary clutter if you're circumventing the whole module system anyway

scrdest replied:
It it, yeah, but even worst-case scenario, this is not on the hot path anywhere (except possibly periodic deepmaint rebuild, which is not even currently used by anything).

I would like to see what API people use more, initially.

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,151 @@
+		var/mob/M = trg
+		if(istype(M))
+			M.Stun(1)
+			M.flash_eyes(intensity = FLASH_PROTECTION_MAJOR * 10, override_blindness_check = TRUE, affect_silicon = TRUE, visual = FALSE, type = /obj/screen/fullscreen/blackout)
```
Nothing major, just a little peeve. As these parameters are quite long, could we newline this for better readability?
```suggestion
			M.flash_eyes(
				intensity = FLASH_PROTECTION_MAJOR * 10, 
				override_blindness_check = TRUE,
				affect_silicon = TRUE,
				visual = FALSE,
				type = /obj/screen/fullscreen/blackout
			)
```

scrdest replied:
Sure, more readability = more good

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/teleporters.dm:
```diff
@@ -0,0 +1,151 @@
+		var/mob/M = trg
+		if(istype(M))
+			M.Stun(1)
+			M.flash_eyes(intensity = FLASH_PROTECTION_MAJOR * 10, override_blindness_check = TRUE, affect_silicon = TRUE, visual = FALSE, type = /obj/screen/fullscreen/blackout)
```
As above

CPU-Blanc commented on maps/deepmaint_wfc/wfc_deepmaint.dm:
```diff
@@ -0,0 +1,25 @@
+/datum/map_template/ruin/deepmaint_wfc
+	name = "Maintrooms"
+	id = "deepmaint_wfc"
+	description = "Somewhere inbetween. How did we get here? How do we leave?"
```
Time to be picky with grammar :D
```suggestion
	description = "Somewhere in-between. How did we get here? How do we leave?"
```

scrdest replied:
I'll change it, but it's a direct copypasta from Bay Deepmaint as-is.

CPU-Blanc reviewed (commented):
Latest batch of smooth-brain comments :D

scrdest committed:
Apply suggestions from code review

Co-authored-by: CPU_Blanc <54823378+CPU-Blanc@users.noreply.github.com>

Commit:
fix: rebase + missing fix from CR

CPU-Blanc commented on code/modules/events/event_container.dm:
```diff
@@ -176,6 +176,7 @@ var/global/list/severity_to_string = list(EVENT_LEVEL_MUNDANE = "Mundane", EVENT
 		new /datum/event_meta(EVENT_LEVEL_MODERATE, "Drone Uprising",						/datum/event/rogue_maint_drones,		25,		list(ASSIGNMENT_ENGINEER = 30)),
 		new /datum/event_meta(EVENT_LEVEL_MODERATE, "Supermatter Power Surge",				/datum/event/power_surge,				100,	list(ASSIGNMENT_ENGINEER = 10)),
 		new /datum/event_meta(EVENT_LEVEL_MODERATE, "Bluespace Drive Instability",			/datum/event/bsd_instability,			50),
+		new /datum/event_meta(EVENT_LEVEL_MODERATE, "DeepmaintRooms",						/datum/event/backrooms_deepmaint,		50),
```
Should this really be a moderate event? I've not poked the event code much so please do correct any misconceptions, but when this event ends, aren't all people stuck in deepmaint just... SOL and trapped forever?

scrdest replied:
They aren't, generally. Even during the event, you can escape. This is not a property of the event, just of Deepmaint itself. 

All standard Deepmaint turfs have a small chance of having a hidden station teleport trait, which itself has a probability of triggering on each crossing.

Why the double-probability? Well, for one, it keeps the number of rolls down, but mostly - this is the main actual _threat_. Not monsters or other hostiles (even though they may rarely be added eventually) - there is a small-moderate chance you get walled in a section with no station teleporters in it.

Even then, if you die, you will get yeeted to regular maint.

CPU-Blanc replied:
Oh. DERP. Just saw that it deletes all the _entrances_. My brain misread this as all teleporters for wfc.

CPU-Blanc commented on code/modules/events/event_container.dm:
```diff
@@ -189,7 +190,8 @@ var/global/list/severity_to_string = list(EVENT_LEVEL_MUNDANE = "Mundane", EVENT
 		new /datum/event_meta(EVENT_LEVEL_MAJOR, "Spider Infestation",				/datum/event/spider_infestation, 	0,	list(ASSIGNMENT_SECURITY = 10), 1),
-		new /datum/event_meta(EVENT_LEVEL_MAJOR, "Bluespace Drive Instability",	/datum/event/bsd_instability,		0,		null, 1),
+		new /datum/event_meta(EVENT_LEVEL_MAJOR, "Bluespace Drive Instability",		/datum/event/bsd_instability,		0,	null, 1),
+		new /datum/event_meta(EVENT_LEVEL_MAJOR, "DeepmaintRooms",					/datum/event/backrooms_deepmaint,	0,	null, 1),
```
Again, seems weird to have this in both event categories where the only difference here is it'll only happens once. Is this an oversight?

scrdest replied:
TBH I was copying BSD Instability above :^)

CPU-Blanc commented on code/modules/urist/modules/wfc/modules/deepmaint/event.dm:
```diff
@@ -0,0 +1,109 @@
+	var/list/spawned_teleporters
+
+
+/datum/event/gravity/setup()
+	endWhen = rand(600, 6000)
```
Seems odd to overload the gravity event here. Error when copying an existing event as a template?
```suggestion
/datum/event/backrooms_deepmaint/setup()
	endWhen = rand(600, 6000)
```

scrdest replied:
Oh, lol, yeah

scrdest committed:
Update code/modules/urist/modules/wfc/modules/deepmaint/event.dm

Co-authored-by: CPU_Blanc <54823378+CPU-Blanc@users.noreply.github.com>

CPU-Blanc reviewed (approved):
LGTM. All my outstanding comments are things that honestly are mostly tweaks that are fine either way. Haven't looked through the DLL Rust repo just yet but as that's just a straight single file drop-in I'm happy to look over that at a later time and get this in for now :+1:

Glloyd committed:
Merge branch 'baymerge-testing' into feature/deepmaint

Glloyd commented:
thanks for taking the lead on this @CPU-Blanc
